lib/nlp/text_normalization.py

```diff
@@ -0,0 +1,196 @@
+
+    pattern = f'(?b)({regex_pattern}){{i<={insert_edits}}}'
+    pattern = re.compile(pattern, flags=_re_flags)
+    matched_format = pattern.search(input_text)
+
+    # Fuzzy matching acts in a non-greedy fashion, hence the following resolution of reverse iterations
+    range_matches = re.finditer(CAPTURE_RANGE_RE, regex_pattern)
+    for match in range_matches:
+        min_range = int(match["minimum"])
+        max_range = int(match["maximum"])
+        for i in range(max_range, min_range - 1, -1):
+            temp_pattern = regex_pattern.replace(match.group(), f'{{{i}}}')
+            pattern = f'(?b)({temp_pattern}){{i<={insert_edits}}}'
+            pattern = re.compile(pattern)
+            matched_format = pattern.search(input_text)
+            if matched_format:
+                regex_pattern = temp_pattern
+                break
+            if i == min_range:
+                return input_text
+
+    if matched_format:
+        if any(matched_format.fuzzy_counts):
+            fuzzy_edits = matched_format.fuzzy_changes[1]  # Insert edits are returned at position 1 in the tuple
+            for corrector, index in enumerate(sorted(fuzzy_edits, reverse=False)):
+                index -= corrector
+                input_text = _omit_character_by_index(input_text, index)
+    return input_text
+
+
+def _omit_character_by_index(text, index) -> str:
+    return text[:index] + text[index + 1:]
+
+
+def resolve_numerals(text) -> str:
+    """
+    Uses NumberDetector to resolve numeric occurrences in text for both English and Hindi.
+    Args:
+        text (str): processed string with numerals and character constants fixed
+    Returns:
+        processed_text (str): modified text
+    """
+    processed_text = text
+    number_detector = NumberDetector('asr_dummy', language='en')
+    # FIXME: Detection fails if text starts with '0' since number detector discards it
+    detected_numerals, original_texts = number_detector.detect_entity(text=text)
+    detected_numerals_hi, original_texts_hi = number_detector.detect_entity(text=text, language='hi')
+    detected_numerals.extend(detected_numerals_hi)
+    original_texts.extend(original_texts_hi)
+    for number, original_text in zip(detected_numerals, original_texts):
+        substitution_reg = re.compile(re.escape(original_text), re.IGNORECASE)
+        processed_text = substitution_reg.sub(number[NUMBER_DETECTION_RETURN_DICT_VALUE], processed_text)
+    return processed_text
+
+
+def resolve_characters(text) -> str:
+    """
+    Uses a dictionary to resolve hindi character occurrences in text to English.
+    Args:
+        text (str): processed string with numerals fixed
+    Returns:
+        processed_text (str): modified text
+    """
+    processed_text = text
+    occurrences = []
+    for char in CHARACTER_CONSTANTS.keys():
+        if char in text:
+            occurrences.append(char)
+    for fragment in sorted(occurrences, key=len):
```

Comment:
Is there a reason we are trying to sort the character occurrences before replacing them?

Reply:
Yes, there might be instances where entities might overlap. 
Eg. "mere untees rupaye hue" -> "tees" should not get detected and changed.

Reply:
That's a good corner case you covered up.